Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,2354 @@
+bool CableSubsystem::isInstanceOf(const Subsystem& s)
+{
+    return Impl::isA(s.getSubsystemGuts());
+}
+
+const CableSubsystem& CableSubsystem::downcast(const Subsystem& s)
+{
+    assert(isInstanceOf(s));
+    return static_cast<const CableSubsystem&>(s);
+}
+
+CableSubsystem& CableSubsystem::updDowncast(Subsystem& s)
+{
+    assert(isInstanceOf(s));
+    return static_cast<CableSubsystem&>(s);
+}
+
+CableSubsystem::CableSubsystem()
+{
+    adoptSubsystemGuts(new Impl());
+}
+
+CableSubsystem::CableSubsystem(MultibodySystem& mbs)
+{
+    adoptSubsystemGuts(new Impl());
+    mbs.adoptSubsystem(*this);
+}
+
+int CableSubsystem::getNumCables() const
+{
+    return getImpl().getNumCables();
+}
+
+const CableSpan& CableSubsystem::getCable(CableSpanIndex ix) const
+{
+    return getImpl().getCable(ix);
+}
+
+CableSpan& CableSubsystem::updCable(CableSpanIndex ix)
+{
+    return updImpl().updCable(ix);
+}
+
+//------------------------------------------------------------------------------
+//                        CABLE SUBSYSTEM IMPL
+//------------------------------------------------------------------------------
+
+const CableSubsystem::Impl& CableSubsystem::getImpl() const
+{
+    return SimTK_DYNAMIC_CAST_DEBUG<const Impl&>(getSubsystemGuts());
+}
+
+CableSubsystem::Impl& CableSubsystem::updImpl()
+{
+    return SimTK_DYNAMIC_CAST_DEBUG<Impl&>(updSubsystemGuts());
+}
+
+int CableSubsystem::Impl::calcDecorativeGeometryAndAppendImpl(
+    const State& state,
+    Stage stage,
+    Array_<DecorativeGeometry>& decorations) const
+{
+    if (stage != Stage::Position)
+        return 0;
+
+    for (const CableSpan& cable : cables) {
+        int returnValue = cable.getImpl().calcDecorativeGeometryAndAppend(
+            state,
+            stage,
+            decorations);
+        if (returnValue != 0) {
+            return returnValue;
+        }
+    }
+    return 0;
+}
+
+int CableSubsystem::Impl::realizeSubsystemTopologyImpl(State& state) const
+{
+    // Briefly allow writing into the Topology cache; after this the
+    // Topology cache is const.
+    Impl* wThis = const_cast<Impl*>(this);
+
+    wThis->realizeTopology(state);
+    for (CableSpanIndex ix(0); ix < cables.size(); ++ix) {
+        CableSpan& path = wThis->updCable(ix);
+        path.updImpl().realizeTopology(state);
+    }
+
+    return 0;
+}
+
+//==============================================================================
+//                                CABLE SPAN
+//==============================================================================
+
+// Allocate a new default-constructed CableSpan::Impl.
+CableSpan::CableSpan() : m_Impl(std::shared_ptr<Impl>(new Impl())) {}
+
+// Invalidate the CableSpan::Impl, and delete the reference. Anyone still
+// holding on to the data (like the subsystem) will know it is invalid, and
+// delete their reference as well.
+CableSpan::~CableSpan()
+{
+    if (m_Impl) {
+        m_Impl->invalidateSubsystemEntry();
+        m_Impl = nullptr;
```

Comment:
Unnecessary - after a dtor is ran, each member variable's destructors are ran in reverse-order, and `std::shared_ptr` will handle its own destruction.